src/dfx-core/src/config/model/dfinity.rs

```diff
@@ -950,42 +951,65 @@ impl Config {
         Ok(None)
     }
 
-    fn from_file(path: &Path) -> Result<Config, StructuredFileError> {
+    fn from_file(
+        path: &Path,
+        dfx_version: &semver::Version,
+    ) -> Result<Config, StructuredFileError> {
         let content = crate::fs::read(path).map_err(ReadJsonFileFailed)?;
-        Config::from_slice(path.to_path_buf(), &content)
+        Config::from_slice(path.to_path_buf(), &content, dfx_version)
     }
 
-    pub fn from_dir(working_dir: &Path) -> Result<Option<Config>, LoadDfxConfigError> {
+    pub fn from_dir(
+        working_dir: &Path,
+        dfx_version: &semver::Version,
+    ) -> Result<Option<Config>, LoadDfxConfigError> {
         let path = Config::resolve_config_path(working_dir)?;
-        path.map(|path| Config::from_file(&path))
+        path.map(|path| Config::from_file(&path, dfx_version))
             .transpose()
             .map_err(LoadFromFileFailed)
     }
 
-    pub fn from_current_dir() -> Result<Option<Config>, LoadDfxConfigError> {
-        Config::from_dir(&std::env::current_dir().map_err(DetermineCurrentWorkingDirFailed)?)
+    pub fn from_current_dir(
+        dfx_version: &semver::Version,
+    ) -> Result<Option<Config>, LoadDfxConfigError> {
+        Config::from_dir(
+            &std::env::current_dir().map_err(DetermineCurrentWorkingDirFailed)?,
+            dfx_version,
+        )
     }
 
-    fn from_slice(path: PathBuf, content: &[u8]) -> Result<Config, StructuredFileError> {
-        let config = serde_json::from_slice(content)
+    fn from_slice(
+        path: PathBuf,
+        content: &[u8],
+        dfx_version: &semver::Version,
+    ) -> Result<Config, StructuredFileError> {
+        let mut json: serde_json::Value = serde_json::from_slice(content)
             .map_err(|e| DeserializeJsonFileFailed(Box::new(path.clone()), e))?;
-        let json = serde_json::from_slice(content)
+        let extension_manager =
+            crate::extension::manager::ExtensionManager::new(dfx_version).unwrap();
+        custom_canister_type::transform_dfx_json_via_extension(&mut json, extension_manager)
+            .unwrap(); // TODO: error handling
```

Comment:
I am not sure what kind of error to use over here

Reply:
Since from_slice will now have new failure modes, we'll have to change the return type.

We can reduce the scope of those failure modes, though.

How about this?

1. Instantiate the ExtensionManager in the caller
 - make EnvironmentImpl::new() create the ExtensionManager.
 - rename EnvironmentImpl::new_extension_manager()` and 

2. Pass a trait rather than the whole ExtensionManager 
- make `trait TransformConfiguration`
   - single method: `transform(&mut json) -> Result<(), TransformConfigurationError>`
   - either pass `Option<TransformConfiguration>` to these methods, or for tests pass an object that is noop for this method

3. Change these methods' error type from StructuredFileError to ReadConfigurationError (possibilities: StructuredFileError, TransformConfigurationError)

This way the error return type of `from_slice` and friends will reflect the errors that can occur, but won't also have to include errors related to instantiating the extension manager.